hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/FileGroupReaderBasedMergeHandle.java

```diff
@@ -20,81 +20,80 @@
-                                             CompactionOperation operation, TaskContextSupplier taskContextSupplier,
-                                             Option<BaseKeyGenerator> keyGeneratorOpt,
-                                             HoodieReaderContext<T> readerContext) {
+@NotThreadSafe
+public class FileGroupReaderBasedMergeHandle<T, I, K, O> extends HoodieMergeHandle<T, I, K, O> {
+  private static final Logger LOG = LoggerFactory.getLogger(FileGroupReaderBasedMergeHandle.class);
+
+  private final HoodieReaderContext<T> readerContext;
+  private final FileSlice fileSlice;
+  private final CompactionOperation operation;
+  private final String maxInstantTime;
+  private HoodieReadStats readStats;
+  private final HoodieRecord.HoodieRecordType recordType;
+
+  public FileGroupReaderBasedMergeHandle(HoodieWriteConfig config, String instantTime, HoodieTable<T, I, K, O> hoodieTable,
+                                         FileSlice fileSlice, CompactionOperation operation, TaskContextSupplier taskContextSupplier,
+                                         HoodieReaderContext<T> readerContext, String maxInstantTime,
+                                         HoodieRecord.HoodieRecordType enginRecordType) {
     super(config, instantTime, operation.getPartitionPath(), operation.getFileId(), hoodieTable, taskContextSupplier);
+    this.maxInstantTime = maxInstantTime;
     this.keyToNewRecords = Collections.emptyMap();
     this.readerContext = readerContext;
-    Option<HoodieBaseFile> baseFileOpt =
-        operation.getBaseFile(config.getBasePath(), operation.getPartitionPath());
-    List<HoodieLogFile> logFiles = operation.getDeltaFileNames().stream().map(p ->
-            new HoodieLogFile(new StoragePath(FSUtils.constructAbsolutePath(
-                config.getBasePath(), operation.getPartitionPath()), p)))
-        .collect(Collectors.toList());
-    this.fileSlice = new FileSlice(
-        operation.getFileGroupId(),
-        operation.getBaseInstantTime(),
-        baseFileOpt.isPresent() ? baseFileOpt.get() : null,
-        logFiles);
+    this.fileSlice = fileSlice;
+    this.operation = operation;
     this.preserveMetadata = true;
-    init(operation, this.partitionPath, baseFileOpt);
-    validateAndSetAndKeyGenProps(keyGeneratorOpt, config.populateMetaFields());
-  }
-
-  private void validateAndSetAndKeyGenProps(Option<BaseKeyGenerator> keyGeneratorOpt, boolean populateMetaFields) {
-    ValidationUtils.checkArgument(populateMetaFields == !keyGeneratorOpt.isPresent());
-    this.keyGeneratorOpt = keyGeneratorOpt;
+    // If the table is a metadata table or the base file is an HFile, we use AVRO record type, otherwise we use the engine record type.
+    this.recordType = hoodieTable.isMetadataTable() || HFILE.getFileExtension().equals(hoodieTable.getBaseFileExtension()) ? HoodieRecord.HoodieRecordType.AVRO : enginRecordType;
```

Comment:
instead of fixing this here, shouldn't we fix 
HoodieCompactor.getEngineRecordType() to return the right value. 
we might have to rename the method though if we fix it

Reply:
sg. if this is the only use of deducing record type, its fine.